zookeeper-server/src/main/java/org/apache/zookeeper/server/NettyServerCnxn.java

```diff
@@ -115,6 +115,9 @@ public void close() {
         // if this is not in cnxns then it's already closed
         if (!factory.cnxns.remove(this)) {
             LOG.debug("cnxns size:{}", factory.cnxns.size());
+            if (channel.isOpen()) {
```

Comment:
Does `CertificateVerifier` take care of setting the `DisconnectReason`? Do we print them somewhere in server logs?

Reply:
It prints "authentication failed" in the server logs.

Reply:
I'm not familiar with the concept of `DisconnectReason`, need to take a look first. There should be quite straightforward to add it to CertificateVerifier if it's any help to you.

Reply:
Oh, sorry, I'm completely wrong. It's already done.
```
                    if (KeeperException.Code.OK != authProvider.handleAuthentication(cnxn, null)) {
                        LOG.error("Authentication failed for session 0x{}", Long.toHexString(cnxn.getSessionId()));
                        cnxn.close(ServerCnxn.DisconnectReason.SASL_AUTH_FAILURE);
                        return;
                    }
```